Crash offline plugin when cancelling download

Click cancel download in the notification when an offline download is ongoing (tap quickly multiple times):

```java
    java.lang.RuntimeException: Unable to start service com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadService@edff33d with Intent { act=com.mapbox.mapboxsdk.plugins.offline.download.cancel flg=0x10000000 cmp=com.mapbox.mapboxsdk.plugins.testapp/com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadService bnds=[1112,688][1392,816] (has extras) }: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.mapbox.mapboxsdk.offline.OfflineRegion.setDownloadState(int)' on a null object reference
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3694)
        at android.app.ActivityThread.-wrap21(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1801)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6938)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void com.mapbox.mapboxsdk.offline.OfflineRegion.setDownloadState(int)' on a null object reference
        at com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadService.cancelDownload(OfflineDownloadService.java:177)
        at com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadService.onResolveCommand(OfflineDownloadService.java:111)
        at com.mapbox.mapboxsdk.plugins.offline.offline.OfflineDownloadService.onStartCommand(OfflineDownloadService.java:87)
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:3677)
        at android.app.ActivityThread.-wrap21(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1801) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6938) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374) 

```

--- a/plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/offline/OfflineDownloadService.java
+++ b/plugin-offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/offline/OfflineDownloadService.java
@@ -110,7 +110,6 @@ public class OfflineDownloadService extends Service {
       createDownload(offlineDownload);
     } else if (OfflineConstants.ACTION_CANCEL_DOWNLOAD.equals(intentAction)) {
       cancelDownload(offlineDownload);
-      stopSelf(offlineDownload.uuid().intValue());
     }
   }
 
@@ -153,8 +152,11 @@ public class OfflineDownloadService extends Service {
     createMapSnapshot(offlineDownload.definition(), new MapSnapshotter.SnapshotReadyCallback() {
       @Override
       public void onSnapshotReady(MapSnapshot snapshot) {
-        notificationBuilder.setLargeIcon(snapshot.getBitmap());
-        notificationManager.notify(offlineDownload.uuid().intValue(), notificationBuilder.build());
+        final int regionId = offlineDownload.uuid().intValue();
+        if (regionLongSparseArray.get(regionId) != null) {
+          notificationBuilder.setLargeIcon(snapshot.getBitmap());
+          notificationManager.notify(regionId, notificationBuilder.build());
+        }
       }
     });
   }
@@ -175,22 +177,34 @@ public class OfflineDownloadService extends Service {
   private void cancelDownload(final OfflineDownloadOptions offlineDownload) {
     int serviceId = offlineDownload.uuid().intValue();
     OfflineRegion offlineRegion = regionLongSparseArray.get(serviceId);
-    offlineRegion.setDownloadState(OfflineRegion.STATE_INACTIVE);
-    offlineRegion.setObserver(null);
-    offlineRegion.delete(new OfflineRegion.OfflineRegionDeleteCallback() {
-      @Override
-      public void onDelete() {
-        // no-op
-      }
+    if (offlineRegion != null) {
+      offlineRegion.setDownloadState(OfflineRegion.STATE_INACTIVE);
+      offlineRegion.setObserver(null);
+      offlineRegion.delete(new OfflineRegion.OfflineRegionDeleteCallback() {
+        @Override
+        public void onDelete() {
+          // no-op
+        }
 
-      @Override
-      public void onError(String error) {
-        OfflineDownloadStateReceiver.dispatchErrorBroadcast(getApplicationContext(), offlineDownload, error);
-      }
-    });
+        @Override
+        public void onError(String error) {
+          OfflineDownloadStateReceiver.dispatchErrorBroadcast(getApplicationContext(), offlineDownload, error);
+        }
+      });
+    }
     OfflineDownloadStateReceiver.dispatchCancelBroadcast(getApplicationContext(), offlineDownload);
-    notificationManager.cancel(serviceId);
-    stopSelf(serviceId);
+    removeOfflineRegion(serviceId);
+  }
+
+  private synchronized void removeOfflineRegion(int regionId) {
+    if (notificationBuilder != null) {
+      notificationManager.cancel(regionId);
+    }
+    regionLongSparseArray.remove(regionId);
+    if (regionLongSparseArray.size() == 0) {
+      stopForeground(true);
+    }
+    stopSelf(regionId);
   }
 
   void launchDownload(final OfflineDownloadOptions offlineDownload, final OfflineRegion offlineRegion) {
@@ -231,14 +245,10 @@ public class OfflineDownloadService extends Service {
    * @since 0.1.0
    */
   void finishDownload(OfflineDownloadOptions offlineDownload, OfflineRegion offlineRegion) {
-    if (notificationBuilder != null) {
-      notificationManager.cancel(offlineDownload.uuid().intValue());
-    }
     OfflineDownloadStateReceiver.dispatchSuccessBroadcast(this, offlineDownload);
     offlineRegion.setDownloadState(OfflineRegion.STATE_INACTIVE);
     offlineRegion.setObserver(null);
-    regionLongSparseArray.remove(offlineDownload.uuid().intValue());
-    stopSelf(offlineDownload.uuid().intValue());
+    removeOfflineRegion(offlineDownload.uuid().intValue());
   }
 
   void progressDownload(OfflineDownloadOptions offlineDownload, OfflineRegionStatus status) {
@@ -248,7 +258,7 @@ public class OfflineDownloadService extends Service {
 
     offlineDownload = offlineDownload.toBuilder().progress(percentage).build();
 
-    if (percentage % 2 == 0) {
+    if (percentage % 2 == 0 && regionLongSparseArray.get(offlineDownload.uuid().intValue()) != null) {
       OfflineDownloadStateReceiver.dispatchProgressChanged(this, offlineDownload, percentage);
       if (notificationBuilder != null) {
         notificationBuilder.setProgress(100, percentage, false);

